scala/scala.bzl

```diff
@@ -100,18 +86,123 @@ touch -t 198001010000 {manifest}
+    "-sbt-interface", ctx.file._sbt_interface_jar.path,
+    "-compiler-interface", ctx.file._compiler_interface_jar.path,
+    "-cp {jars}",
+    "-d", tmp_out_dir.path,
+  ]
+  flags = " ".join(flags)
+  flags = flags.format(
+      out=ctx.outputs.jar.path,
+      jars=":".join([j.path for j in jars]))
+  work_unit_args = ctx.new_file(ctx.configuration.bin_dir, ctx.label.name + "_args")
+  ctx.file_action(output = work_unit_args, content=flags)
+
+  # Generate the "@"-file containing the command-line args for the unit of work.
+  argfile = ctx.new_file(ctx.configuration.bin_dir, "worker_input")
+  argfile_contents = "\n".join(["-argfile", work_unit_args.path] + [f.path for f in ctx.files.srcs])
+  ctx.file_action(output=argfile, content=argfile_contents)
+
+  # Classpath for the compiler/worker itself, these are not the compile time dependencies.
+  classpath_jars = [
+      ctx.file._compiler_interface_jar,
+      ctx.file._incremental_compiler_jar,
+      ctx.file._scala_compiler_jar,
+      ctx.file._scala_library_jar,
+      ctx.file._scala_reflect_jar,
+      ctx.file._sbt_interface_jar,
+      ctx.file._zinc,
+      ctx.file._zinc_compiler_jar,
+      ctx.file._nailgun_server_jar,
+  ]
+  compiler_classpath = ":".join([f.path for f in classpath_jars])
+
+  ctx.action(
+      inputs=list(jars) + ctx.files.srcs + [ctx.outputs.manifest, argfile, work_unit_args] + classpath_jars,
+      outputs=[tmp_out_dir],
+      executable=worker,
+      progress_message="Zinc Worker: %s" % ctx.label.name,
+      mnemonic="Scala",
+      arguments=ctx.attr.worker_args + [compiler_classpath] + ["@" + argfile.path],
+  )
+
+  cmd = """
+set -e
+# Make jar file deterministic by setting the timestamp of files
+find {tmp_out} | xargs touch -t 198001010000
+touch -t 198001010000 {manifest}
+jar cmf {manifest} {out} -C {tmp_out} .
+""" + _get_res_cmd(ctx)
+  cmd = cmd.format(
+      tmp_out=tmp_out_dir.path,
+      out=ctx.outputs.jar.path,
+      manifest=ctx.outputs.manifest.path)
+
+  ctx.action(
+      inputs=[tmp_out_dir, ctx.outputs.manifest],
+      outputs=[ctx.outputs.jar],
+      command=cmd,
+      progress_message="Building Jar: %s" % ctx.label)
+
+
+def _get_res_cmd(ctx):
+  res_cmd = ""
+  for f in ctx.files.resources:
+    c_dir, res_path = _adjust_resources_path(f.path)
+    change_dir = "-C " + c_dir if c_dir else ""
+    res_cmd = "\n{jar} uf {out} " + change_dir + " " + res_path
+    res_cmd = res_cmd.format(
+      out=ctx.outputs.jar.path,
+      jar=ctx.file._jar.path,)
+  return res_cmd
+
+def _build_ijar(ctx):
+  ijar_cmd = """
+    set -e
+    {ijar} {out} {ijar_out}
+  """.format(
+    ijar=ctx.file._ijar.path,
+    out=ctx.outputs.jar.path,
+    ijar_out=ctx.outputs.ijar.path)
+
+  ctx.action(
+    inputs=[ctx.outputs.jar, ctx.file._ijar],
+    outputs=[ctx.outputs.ijar],
+    command=ijar_cmd,
+    progress_message="scala ijar %s" % ctx.label,)
+
+
+def _compile(ctx, jars, buildijar, usezinc):
```

Comment:
Refactored this a bit so that it can be reused for both scalac and zinc invocations